Source/WebCore/page/ElementTargetingController.cpp

```diff
@@ -860,6 +861,129 @@ Vector<TargetedElementInfo> ElementTargetingController::findTargets(TargetedElem
+        return { };
+    }
+
+    RefPtr view = mainFrame->view();
+    if (!view) {
+        ASSERT_NOT_REACHED();
+        return { };
+    }
+
+    const auto viewportRect = view->unobscuredContentRect();
+    const auto hitTestRegionWidth = minimumSize.width();
+    const auto hitTestRegionHeight = minimumSize.height();
+
+    Vector<Vector<TargetedElementInfo>> targetsList;
+    for (auto x = viewportRect.x() + std::floor(hitTestRegionWidth / 2); x < viewportRect.maxX(); x += hitTestRegionWidth) {
+        for (auto y = viewportRect.y() + std::floor(hitTestRegionHeight / 2); y < viewportRect.maxY(); y += hitTestRegionHeight) {
+            const FloatPoint hitTestPoint = FloatPoint(x, y);
+            auto [nodes, innerElement] = findNodes(hitTestPoint, true);
+            if (nodes.isEmpty())
+                continue;
+
+            const auto targets = extractTargets(WTFMove(nodes), WTFMove(innerElement), false);
+            targetsList.append(targets);
+        }
+    }
+
+    HashMap<WebCore::ElementIdentifier, HashSet<WebCore::ElementIdentifier>> elementIDToOccludedElementIDs;
+    HashMap<WebCore::ElementIdentifier, Vector<TargetedElementInfo>> elementIDToTargets;
+    for (auto& targets : targetsList) {
+        if (targets.size() <= 0)
+            continue;
+
+        const auto topElementID = targets.first().elementIdentifier;
+        HashSet<WebCore::ElementIdentifier> occludedElementIDsToInsert;
+        bool isFirstElement = true;
+        for (auto& target : targets) {
+            if (isFirstElement) {
+                isFirstElement = false;
+                continue;
+            }
+
+            occludedElementIDsToInsert.add(target.elementIdentifier);
+        }
+
+        if (elementIDToTargets.contains(topElementID)) {
+            const auto storedIDsSet = elementIDToOccludedElementIDs.get(topElementID);
+            auto storedTargets = elementIDToTargets.get(topElementID);
+            for (auto& target : targets) {
+                if (target.elementIdentifier != topElementID && !storedIDsSet.contains(target.elementIdentifier))
+                    storedTargets.append(target);
+            }
+
+            elementIDToTargets.set(topElementID, storedTargets);
+        } else
+            elementIDToTargets.set(topElementID, targets);
+
+        if (elementIDToOccludedElementIDs.contains(topElementID))
+            elementIDToOccludedElementIDs.set(topElementID,  elementIDToOccludedElementIDs.get(topElementID).unionWith(occludedElementIDsToInsert));
+        else
+            elementIDToOccludedElementIDs.set(topElementID, occludedElementIDsToInsert);
+    }
+
+    const auto topologicallySortedIDs = topologicallySortElements(elementIDToOccludedElementIDs);
+    Vector<Vector<TargetedElementInfo>> topologicallySortedTargetsList;
+    for (auto& elementID : topologicallySortedIDs)
+        topologicallySortedTargetsList.append(elementIDToTargets.get(elementID));
+
+    return topologicallySortedTargetsList;
```

Comment:
Optional; this plumbing can be wrapped into a `map` operation. I don't think it harms legibility.

```suggestion
    return topologicallySortElements(elementIDToOccludedElementIDs).map([& elementIDToTargets](const auto& elementIdentifier) {
        return elementIDToTargets.get(elementIdentifier);
    });
```